File: R/apply_frozen_model.R
full_args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", full_args, value = TRUE)
script_dir <- if (length(file_arg)) dirname(normalizePath(sub("^--file=", "", file_arg[[1]]))) else getwd()
source(file.path(script_dir, "common.R"))

args <- parse_cli_args()
model_path <- arg_value(args, "model", required = TRUE)
data_path <- arg_value(args, "data", required = TRUE)
id_col <- arg_value(args, "id-column", default = NULL)
out_path <- arg_value(args, "output", default = "transfer_predictions.csv")

artifact <- readRDS(model_path)
dat <- read.csv(data_path, check.names = FALSE, stringsAsFactors = FALSE)
missing <- setdiff(artifact$selected_features, colnames(dat))
if (length(missing)) stop("Transfer data are missing required features: ", paste(head(missing, 20), collapse = ", "))

x <- dat[, artifact$selected_features, drop = FALSE]
xp <- apply_preprocessor(x, artifact$preprocessor)
prob <- predict_model_prob(artifact$model_name, artifact$model, xp)
pred <- ifelse(prob >= 0.5, artifact$positive_label, artifact$negative_label)
out <- data.frame(predicted_probability = prob, predicted_class = pred, check.names = FALSE)
if (!is.null(id_col)) {
  if (!id_col %in% colnames(dat)) stop("ID column absent from transfer data: ", id_col)
  out <- cbind(dat[, id_col, drop = FALSE], out)
}
write.csv(out, out_path, row.names = FALSE)

File: python/apply_frozen_model.py
from __future__ import annotations

import argparse
import joblib
import pandas as pd


def main():
    parser = argparse.ArgumentParser(description="Apply a frozen Python model to a transfer/validation cohort.")
    parser.add_argument("--model", required=True)
    parser.add_argument("--data", required=True)
    parser.add_argument("--id-column", default=None)
    parser.add_argument("--output", default="transfer_predictions.csv")
    args = parser.parse_args()

    artifact = joblib.load(args.model)
    df = pd.read_csv(args.data)
    selected = artifact["selected_features"]
    missing = [f for f in selected if f not in df.columns]
    if missing:
        raise ValueError(f"Transfer data are missing {len(missing)} required features: " + ", ".join(missing[:20]))

    prob = artifact["pipeline"].predict_proba(df[selected])[:, 1]
    pred01 = (prob >= 0.5).astype(int)
    pred_label = [artifact["positive_label"] if p == 1 else artifact["negative_label"] for p in pred01]
    out = pd.DataFrame({"predicted_probability": prob, "predicted_class": pred_label})
    if args.id_column:
        if args.id_column not in df.columns:
            raise ValueError(f"ID column '{args.id_column}' is absent from transfer data.")
        out.insert(0, args.id_column, df[args.id_column].values)
    out.to_csv(args.output, index=False)


if __name__ == "__main__":
    main()
